src/Middleware/CompressionGzipMiddleware.php

```diff
@@ -0,0 +1,43 @@
+<?php
+
+namespace React\Http\Middleware;
+
+use Psr\Http\Message\ServerRequestInterface;
+use React\Http\Response;
+
+final class CompressionGzipMiddleware
+{
+
+    private $compressionLevel = -1;
+
+    public function __construct($level = -1)
+    {
+        $this->compressionLevel = $level;
+    }
+
+    public function __invoke(ServerRequestInterface $request, $next)
+    {
+        if (!$request->hasHeader('Accept-Encoding')) {
+            return $next($request);
+        }
+
+        if (stristr($request->getHeaderLine('Accept-Encoding'), 'gzip') === false) {
+            return $next($request);
+        }
+
+        $level = $this->compressionLevel;
+        return $next($request)->then(function (Response $response) use ($request, $next, $level) {
+          if ($response->hasHeader('Content-Encoding')) {
+              return $response;
+          }
+
+          $content = $response->getBody()->getContents();
```

Comment:
This will cast the response body to a string. In case, the body is a `HttpBodyStream`, that [doesn't work](https://github.com/reactphp/http/blob/master/src/HttpBodyStream.php#L143).

You probably need to decorate the stream and compress on the fly, eg. using https://github.com/clue/php-zlib-react.